backend/src/main/java/ai/verta/modeldb/lineage/LineageDAORdbImpl.java

```diff
@@ -30,36 +49,241 @@
   public LineageDAORdbImpl() {}
 
   @Override
-  public Response addLineage(AddLineage addLineage, IsExistsPredicate isExistsPredicate)
-      throws ModelDBException {
+  public Response addLineage(
+      AddLineage addLineage,
+      ExistsCheckConsumer existsCheckConsumer,
+      BlobHashInCommitFunction blobHashInCommitFunction)
+      throws ModelDBException, InvalidProtocolBufferException, NoSuchAlgorithmException {
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
       session.beginTransaction();
+      Long id;
+      LineageElementEntity lineageElementEntity;
+      if (addLineage.getId() != 0) {
+        id = addLineage.getId();
+        lineageElementEntity = session.get(LineageElementEntity.class, addLineage.getId());
+      } else {
+        id = null;
+        lineageElementEntity = null;
+      }
+      if (lineageElementEntity == null) {
+        lineageElementEntity = new LineageElementEntity(id);
+        session.save(lineageElementEntity);
+        id = lineageElementEntity.getId();
+      }
       validate(addLineage.getInputList(), addLineage.getOutputList());
-      validateExistence(
-          addLineage.getInputList(), addLineage.getOutputList(), isExistsPredicate, session);
+      List<LineageEntry> lineageEntries = new LinkedList<>(addLineage.getInputList());
+      lineageEntries.addAll(addLineage.getOutputList());
+      existsCheckConsumer.test(session, lineageEntries);
       for (LineageEntry input : addLineage.getInputList()) {
-        for (LineageEntry output : addLineage.getOutputList()) {
-          addLineage(session, input, output);
-        }
+        addLineage(session, input, id, CONNECTION_TYPE_INPUT, blobHashInCommitFunction);
+      }
+      for (LineageEntry output : addLineage.getOutputList()) {
+        addLineage(session, output, id, CONNECTION_TYPE_OUTPUT, blobHashInCommitFunction);
       }
       session.getTransaction().commit();
+      return AddLineage.Response.newBuilder().setId(lineageElementEntity.getId()).build();
     }
-    return AddLineage.Response.newBuilder().setStatus(true).build();
   }
 
   @Override
-  public DeleteLineage.Response deleteLineage(DeleteLineage deleteLineage) throws ModelDBException {
+  public DeleteLineage.Response deleteLineage(
+      DeleteLineage deleteLineage, BlobHashInCommitFunction blobHashInCommitFunction)
+      throws ModelDBException {
+    boolean status;
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
       session.beginTransaction();
       validate(deleteLineage.getInputList(), deleteLineage.getOutputList());
+      long id = deleteLineage.getId();
+      if (id == 0) {
```

Comment:
Make sure `0` isn't a value that the database will return to you in the creation of the first instance.

Reply:
fixed by adding option to liquibase